src/main/java/picard/sam/markduplicates/UmiUtil.java

```diff
@@ -32,12 +35,111 @@
+     * UMI from a bottom strand can be normalized to be identical to the read from its corresponding top strand
+     * by swapping the content of the UMI around the "-" found in duplex UMIs.  For example, a bottom strand
+     * duplex UMI reading ATC-CGG when top-strand normalized will read CGG-ATC.
+     *
+     * @param record SAM record to retrieve UMI from.
+     * @param umiTag The tag used in the bam file that designates the UMI, null returns null
+     * @return Normalized Duplex UMI.  If the UMI isn't duplex, it returns the UMI unaltered.
+     */
+    static String getTopStrandNormalizedUmi(final SAMRecord record, final String umiTag, final boolean duplexUmi) {
+        if (umiTag == null) {
+            return null;
+        }
+
+        final String umi = record.getStringAttribute(umiTag);
+
+        if (umi == null) {
+            return null;
+        }
+        if (!umi.matches("^[ATCGNatcgn-]*$")) {
+            throw new PicardException("UMI found with illegal characters.  UMIs must match the regular expression ^[ATCGNatcgn-]*$.");
+        }
+
+        if (duplexUmi) {
+            final String[] split = umi.split(DUPLEX_UMI_DELIMITER);
+            if (split.length != 2) {
+                throw new PicardException("Duplex UMIs must be of the form X-Y where X and Y are equal length UMIs, for example AT-GA.  Found UMI, " + umi);
+            }
+
+            if (isTopStrand(record)) {
+                return split[0] + DUPLEX_UMI_DELIMITER + split[1];
+            } else {
+                return split[1] + DUPLEX_UMI_DELIMITER + split[0];
+            }
+        }
+        return umi;
+    }
+
+    /**
+     * Determines if the read represented by a SAM record belongs to the top or bottom strand.
+     * Top strand is defined as having a read 1 unclipped 5' coordinate
+     * less than the read 2 unclipped 5' coordinate.  If a read is unmapped
+     * it is considered to have an unclipped 5' coordinate of 0.
+     * @param rec Record to determine top or bottom strand
+     * @return Top or bottom strand, true (top), false (bottom).
+     */
+    static boolean isTopStrand(final SAMRecord rec) {
+
+        final int read5PrimeStart = (rec.getReadNegativeStrandFlag()) ? rec.getUnclippedEnd() : rec.getUnclippedStart();
+        final int mate5PrimeStart = (rec.getMateNegativeStrandFlag()) ? SAMUtils.getMateUnclippedEnd(rec) : SAMUtils.getMateUnclippedStart(rec);
+        return rec.getFirstOfPairFlag() == (read5PrimeStart < mate5PrimeStart);
+    }
+
+    /**
+     * Creates a string that uniquely identifies a molecule using its UMI and alignment start position
+     * @param rec SAMRecord to create molecular identifier of
+     * @param umiTag Tag that contains the UMI record
+     * @return String that uniquely identifies a fragment
+     */
+    static String molecularIdentifierString(final SAMRecord rec, final String umiTag) {
```

Comment:
I opted to remove this function altogether, and use setMolecularIdentifier instead because I think that is clearer.

setMolecularIdentifier can be used for either duplex or simplex UMIs, and it behaves differently depending on whether or not duplexUmis is set to true/false, so I don't think it should have "Duplex" in it's name.